lib/mix/lib/mix/dep.ex

```diff
@@ -117,18 +117,42 @@ defmodule Mix.Dep do
     end
   end
 
+  @child_keep_opts [:optional, :runtime]
+
   defp load_and_cache(_config, top, top, env) do
     converge(env: env)
   end
 
   defp load_and_cache(config, _top, bottom, _env) do
     {_, deps} = Mix.ProjectStack.read_cache({:cached_deps, bottom})
     app = Keyword.fetch!(config, :app)
-    top_level = for dep <- deps, dep.app == app, child <- dep.deps, do: child.app
-
     seen = populate_seen(MapSet.new(), [app])
     children = get_deps(deps, tl(Enum.uniq(get_children(deps, seen, [app]))))
-    Enum.map(children, &%{&1 | top_level: &1.app in top_level})
+
+    top_level =
+      for dep <- deps,
+          dep.app == app,
+          child <- dep.deps,
+          do: {child.app, Keyword.take(child.opts, @child_keep_opts)},
+          into: %{}
+
+    Enum.map(children, fn %{app: app} = dep ->
+      case top_level do
+        %{^app => child_opts} ->
+          # Only top level dependencies can be optional.
+          # Any non-top level dependency that is optional and
+          # is still available means it has been fulfilled.
```

Comment:
The comment needs to be updated regarding `:runtime`.

Reply:
Done!